tests/tensor/test_subtensor.py

```diff
@@ -2439,25 +2446,86 @@ def test_AdvancedSubtensor_bool(self):
         n = dmatrix()
         n_val = np.arange(6).reshape((2, 3))
 
-        # infer_shape is not implemented, but it should not crash
+        # This case triggers a runtime broadcasting between the nonzero() shapes
+        with pytest.raises(AssertionError, match="hello"):
```

Comment:
It also showed up here. I am unsure about this. I think we either allow runtime broadcasting for the shape graph or we don't do shape inference for the cases where we need to go through the `nonzero`.

Because the Op is happy to broadcast the indexes and won't raise, so the shape shouldn't either. The gradient issue is not a thing here because those inputs are Disconnected.

Reply:
I'm not sure I really understand yet, but...
We are talking about cases like this, right?
```
np.ones((3, 3))[np.zeros((1, 5), dtype=int), np.zeros((7, 1), dtype=int)]
```
which is the same as
```
np.ones((3, 3))[np.zeros((7, 5), dtype=int), np.zeros((7, 5), dtype=int)]
```
Right now, I guess, this broadcasts dynamically.
And the gradients aren't really a problem here, because the corresponding grad op is `AdvancedIndexInc`, which broadcasts the same way.

So I think I agree that gradient-wise this doesn't really seem to be an issue. And not allowing dynamic broadcasting here would (I think) be a breaking change.

So far so good, I'm just a bit scared that this dynamic broadcasting might make it really tricky to write a good implementation of `AdvancedSubtensor` et al. We never had one in the c backend I think, the numba backend doesn't support this (because numba itself doesn't), and the jax backend doesn't really support any dynamic shapes either.

Maybe the reason that nobody ever wrote a good implementation of this is that this is hard to do with dynamic broadcasting?

Reply:
I think the reason is that all these flavors of advanced indexing are just too complex, broadcasting or not. Theano (and us) only have a Python implementation for the general case.

We can always restrict efficient implementations to a subset of cases like the AdvancedSubtensor1 does. In this case we probably won't specialize boolean indexes + other advanced indexes that require broadcast nonzero with those.

Also numpy itself is pivoting towards a more reasonable "outer indexing" by default: https://numpy.org/neps/nep-0021-advanced-indexing.html

I've seen some libraries (xarray, numba) are just skipping ahead to that instead of implementing all these edge cases (like the advanced groups being pushed to the front when they are non contiguous)

On a separate note, I checked numba and @kc611 has implemented some support for some forms of advanced indexing so we may be able to support more than what we did so far in Numba: https://github.com/numba/numba/issues/8616

Reply:
I reintroduced the ability to do `runtime_broadcast` in the `broadcast_shape_iter` as an optional flag.  
This is only used for computing the shape of `AdvancedIndexing`. If we disallow dynamic broadcasting of the multidimensional indexes in the future, than we won't need to worry about it for shape inference either.

I think that decision should be done on a separate PR.